### PR TITLE
Add roundtrip, defaulting, upgrading and validation unit tests for the kubeadm API types

### LINKED ISSUE #841: We need roundtrip, defaulting and validation unit tests between API versions
As part of https://github.com/kubernetes/community/pull/2131, we have concluded we need roundtrip tests between API versions for our internal and external APIs. This combined with only a centralized place in the codebase that handles serialize/deserialize (`cmd/kubeadm/app/util/marshal.go` and `cmd/kubeadm/app/util/config/*.go`), will get us in a good place so these tests are virtually e2e test regarding how kubeadm responds to different configuration files given to it.

This is a priority for v1.11

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:
Follows up from https://github.com/kubernetes/kubernetes/pull/63799, as well as net-new unit testing for our serialization/deserialization package. This tests our API machinery pretty much end to end.

This is more important now given we now support two external types: https://github.com/kubernetes/kubernetes/pull/63788

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of kubernetes/community#2131
Fixes: https://github.com/kubernetes/kubeadm/issues/841

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```
@kubernetes/sig-cluster-lifecycle-pr-reviews @liztio

### CODE REVIEW

File: cmd/kubeadm/app/util/config/masterconfig_test.go [hunk 1]
```diff
@@ -14,5 +14,124 @@ See the License for the specific language governing permissions and
...
+		},
+		{ // v1alpha1 (faulty) -> internal -> v1alpha1
+			in:           v1alpha1WithoutTypeMetaYAML,
+			out:          v1alpha1YAML,
+			groupVersion: v1alpha1.SchemeGroupVersion,
+		},
+	}
+
+	for _, rt := range tests {
+		err := testRoundTrip(rt.in, rt.out, rt.groupVersion)
```
Comment (liztio) [inline cmd/kubeadm/app/util/config/masterconfig_test.go]: use `t.Run` for these, and give them names. Makes it easier to debug failures.

File: cmd/kubeadm/app/util/config/masterconfig_test.go [hunk 2]
```diff
@@ -14,5 +14,124 @@ See the License for the specific language governing permissions and
...
+	if !bytes.Equal(expected, actual) {
+		return fmt.Errorf("the expected and actual output differs.\n\tin: %s\n\tout: %s\n\tgroupversion: %s\n\texpected: %s\n\tactual: %s\n",
+			in, out, gv.String(), expected, actual)
+	}
+	return nil
+}
+
+// TestDefaulting tests reading one file that has only a subset of the fields populated, loading it using ConfigFileAndDefaultsToInternalConfig,
+// and then marshals the internal object to the expected groupVersion
+func TestDefaulting(t *testing.T) {
```
Comment (liztio) [inline cmd/kubeadm/app/util/config/masterconfig_test.go]: why is this a different test to the above? it looks like the body is the same.

File: cmd/kubeadm/app/util/config/testdata/conversion/internal.yaml [hunk 3]
```diff
@@ -0,0 +1,161 @@
+API:
```
Comment (liztio) [inline cmd/kubeadm/app/util/config/testdata/conversion/internal.yaml]: do these need to be fully populated? It makes it difficult to see what's actually being tested.

File: cmd/kubeadm/app/util/config/masterconfig_test.go [hunk 4]
```diff
@@ -14,5 +14,124 @@ See the License for the specific language governing permissions and
...
+	if err != nil {
+		return fmt.Errorf("couldn't unmarshal test data: %v", err)
+	}
+
+	actual, err := kubeadmutil.MarshalToYamlForCodecs(internalcfg, gv, scheme.Codecs)
+	if err != nil {
+		return fmt.Errorf("couldn't marshal internal object: %v", err)
+	}
+
+	if !bytes.Equal(expected, actual) {
```
Comment (liztio) [inline cmd/kubeadm/app/util/config/masterconfig_test.go]: maybe use something like https://godoc.org/github.com/kr/pretty#Diff so the diffs are easier to see?

**Review by liztio — CHANGES_REQUESTED** [review]
These look really good! Just a couple quality-of-life changes to make it easier to maintain

File: cmd/kubeadm/app/util/config/masterconfig_test.go (hunk 4 above)
Comment (luxas) [reply cmd/kubeadm/app/util/config/masterconfig_test.go]: An other time, don't think I have time to make that change currently. Please file a generic issue, as we have this in many unit tests that do this

File: cmd/kubeadm/app/util/config/testdata/conversion/internal.yaml (hunk 3 above)
Comment (luxas) [reply cmd/kubeadm/app/util/config/testdata/conversion/internal.yaml]: Yes, these are roundtrip tests, so the same data is everywhere, but we test conversion and make sure there is no data loss etc.

**Comment by luxas** [comment]
@liztio updated. Now also includes v1alpha2 testing

**Review by timothysc — APPROVED** [review]
/lgtm

**Comment by neolit123** [comment]
```
verify govet 2m50s
make verify WHAT=govet
go build k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1/test: no non-test Go files in /go/src/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1/test
make[1]: *** [vet] Error 1
```

looks like the `go vet` in that toolchain is missing this patch:
https://go-review.googlesource.com/c/go/+/87636

possible fixes:
- move the tests into the `v1alpha1` package
- disable `go vet` for the `test` package
- add dummy, non-test file in the `test` package.

**Comment by luxas** [comment]
Rebased, readding LGTM

**Pull request merged**